File: js/clusterGraph.js
import * as utils from "./utils.js";
import * as gc from "./gc.js";
import { BuildSnnGraphResults } from "./buildSnnGraph.js";

/**
 * Wrapper around multi-level clustering results on the Wasm heap, produced by {@linkcode clusterGraph}.
 * @hideconstructor
 */
export class ClusterMultilevelResults {
    #id;
    #results;

    constructor(id, raw) {
        this.#id = id;
        this.#results = raw;
    }

    /**
     * @return {number} Number of levels in the results.
     */
    numberOfLevels() {
        return this.#results.num_levels();
    }

    /**
     * @return {number} Level with the lowest modularity.
     */
    bestLevel() {
        return this.#results.best_level();
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {?number} [options.level=null] - The clustering level for which to obtain the modularity.
     * Defaults to the best clustering level from {@linkcode ClusterMultilevelResults#bestLevel bestLevel}.
     *
     * @return {number} The modularity at the specified level.
     */
    modularity(options = {}) {
        let { level = null, ...others } = options;
        utils.checkOtherOptions(others);
        if (level == null) {
            return this.#results.best_modularity();
        } else {
            return this.#results.modularity(level);
        }
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {?number} [options.level=null] - The clustering level for which to obtain the cluster membership.
     * Defaults to the best clustering level from {@linkcode ClusterMultilevelResults#bestLevel bestLevel}.
     * @param {boolean} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     * Only used if `level=null`, otherwise a copy is always made.
     *
     * @return {Int32Array|Int32WasmArray} Array containing the cluster membership for each cell.
     */
    membership(options = {}) {
        let { level = null, copy = true, ...others } = options;
        utils.checkOtherOptions(others);
        if (level == null) {
            return utils.possibleCopy(this.#results.best_membership(), copy);
        } else {
            // We must take a copy as this is transient memory.
            return this.#results.membership(level).slice();
        }
    }

    /**
     * @return Frees the memory allocated on the Wasm heap for this object.
     * This invalidates this object and all references to it.
     */
    free() {
        if (this.#results !== null) {
            gc.release(this.#id);
            this.#results = null;
        }
        return;
    }
}

/**
 * Wrapper around the walktrap clustering results on the Wasm heap, produced by {@linkcode clusterGraph}.
 * @hideconstructor
 */
export class ClusterWalktrapResults {
    #id;
    #results;

    constructor(id, raw) {
        this.#id = id;
        this.#results = raw;
    }

    /**
     * @return {number} Number of merge steps used by the Walktrap algorithm.
     */
    numberOfMergeSteps() {
        return this.#results.num_merge_steps();
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {?number} [options.at=null] - Index at which to extract the modularity.
     * This can be any value from 0 to {@linkcode ClusterWalktrapResults#numberOfMergeSteps numberOfMergeSteps} plus 1.
     * Set to `null` to obtain the largest modularity across all merge steps.
     * @return {number} The modularity at the specified merge step, or the maximum modularity across all merge steps.
     */
    modularity(options = {}) {
        let { at = null, ...others } = options;
        utils.checkOtherOptions(others);
        if (at === null) {
            return this.#results.best_modularity();
        } else {
            return this.#results.modularity(at);
        }
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean|string} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     * @return {Int32Array|Int32WasmArray} Array containing the cluster membership for each cell.
     */
    membership(options = {}) {
        const { copy = true, ...others } = options;
        utils.checkOtherOptions(others);
        return utils.possibleCopy(this.#results.membership(), copy);
    }

    /**
     * @return Frees the memory allocated on the Wasm heap for this object.
     * This invalidates this object and all references to it.
     */
    free() {
        if (this.#results !== null) {
            gc.release(this.#id);
            this.#results = null;
        }
        return;
    }
}

/**
 * Wrapper around the Leiden clustering results on the Wasm heap, produced by {@linkcode clusterGraph}.
 * @hideconstructor
 */
export class ClusterLeidenResults {
    #id;
    #results;

    constructor(id, raw, filled = true) {
        this.#id = id;
        this.#results = raw;
        return;
    }

    /**
     * @return {number} The quality of the Leiden clustering.
     *
     * Note that Leiden's quality score is technically a different measure from modularity.
     * Nonetheless, we use `modularity` for consistency with the other SNN clustering result classes.
     */
    modularity() {
        return this.#results.modularity();
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean|string} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     *
     * @return {(Int32Array|Int32WasmArray)} Array containing the cluster membership for each cell.
     */
    membership(options = {}) {
        const { copy = true, ...others } = options;
        utils.checkOtherOptions(others);
        return utils.possibleCopy(this.#results.membership(), copy);
    }

    /**
     * @return Frees the memory allocated on the Wasm heap for this object.
     * This invalidates this object and all references to it.
     */
    free() {
        if (this.#results !== null) {
            gc.release(this.#id);
            this.#results = null;
        }
        return;
    }
}

/**
 * Cluster cells using community detection on the SNN graph.
 *
 * @param {BuildSnnGraphResults} x - The shared nearest neighbor graph constructed by {@linkcode buildSnnGraph}.
 * @param {object} [options={}] - Optional parameters.
 * @param {string} [options.method="multilevel"] - Community detection method to use.
 * This should be one of `"multilevel"`, `"walktrap"` or `"leiden"`.
 * @param {number} [options.multiLevelResolution=1] - The resolution of the multi-level clustering, when `method = "multilevel"`.
 * Larger values result in more fine-grained clusters.
 * @param {number} [options.leidenResolution=1] - The resolution of the Leiden clustering, when `method = "leiden"`.
 * Larger values result in more fine-grained clusters.
 * @param {boolean} [options.leidenObjective="cpm"] - Objective function to use when `method = "leiden"`.
 * This can be the Constant-Potts Model (`cpm`), the graph modularity (`modularity`), or the Erdős-Rényi model (`er`).
 * Set to `modularity` to get an interpretation of the resolution on par with that of `method = "multilevel"`.
 * @param {number} [options.walktrapSteps=4] - Number of steps for the Walktrap algorithm, when `method = "walktrap"`.
 *
 * @return {ClusterMultiLevelResults|ClusterWalktrapResults|ClusterLeidenResults} Object containing the clustering results.
 * The class of this object depends on the choice of `method`.
 */
export function clusterGraph(x, options = {}) {
    const { 
        method = "multilevel", 
        multiLevelResolution = 1, 
        leidenResolution = 1, 
        leidenObjective = "cpm",
        leidenModularityObjective = false, // For back-compatibility only.
        walktrapSteps = 4,
        ...others
    } = options;
    utils.checkOtherOptions(others);

    var output;
    try {
        if (method == "multilevel") {
            output = gc.call(
                module => module.cluster_multilevel(x.graph, multiLevelResolution),
                ClusterMultilevelResults
            );
        } else if (method == "walktrap") {
            output = gc.call(
                module => module.cluster_walktrap(x.graph, walktrapSteps),
                ClusterWalktrapResults
            );
        } else if (method == "leiden") {
            if (leidenModularityObjective) { // for back-compatibility.
                leidenObjective = "modularity";
            }
            output = gc.call(
                module => module.cluster_leiden(x.graph, leidenResolution, leidenObjective),
                ClusterLeidenResults
            );
        } else {
            throw new Error("unknown method '" + method + "'")
        }
    } catch (e) {
        utils.free(output);
        throw e;
    }

    return output;
}

File: src/cluster_graph.cpp
#include <emscripten/bind.h>

#include <vector>
#include <algorithm>
#include <cstdint>

#include "build_snn_graph.h"
#include "utils.h"

#include "scran_graph_cluster/scran_graph_cluster.hpp"
#include "sanisizer/sanisizer.hpp"

class ClusterMultilevelResult {
    typedef scran_graph_cluster::ClusterMultilevelResults Store;

    Store my_store;
    std::size_t my_best = 0;
    raiigraph::IntegerVector my_buffer;

public:
    ClusterMultilevelResult(scran_graph_cluster::ClusterMultilevelResults store) : my_store(std::move(store)) {
        if (store.modularity.size()) {
            sanisizer::can_ptrdiff<I<decltype(store.modularity.begin())> >(store.modularity.size());
            my_best = std::max_element(my_store.modularity.begin(), my_store.modularity.end()) - my_store.modularity.begin();
        }
    }

public:
    JsFakeInt num_levels() const {
        return int2js(my_store.modularity.size());
    }

    JsFakeInt best_level() const {
        return int2js(my_best);
    }

    double modularity(JsFakeInt i_raw) const {
        return my_store.modularity[js2int<std::size_t>(i_raw)];
    }

    emscripten::val membership(JsFakeInt i_raw) {
        sanisizer::resize(my_buffer, my_store.levels.nrow());
        auto row = my_store.levels.row(js2int<std::size_t>(i_raw));
        std::copy(row.begin(), row.end(), my_buffer.begin());
        return emscripten::val(emscripten::typed_memory_view(my_buffer.size(), my_buffer.data()));
    }

    double best_modularity() const {
        return my_store.modularity[my_best];
    }

    emscripten::val best_membership() const {
        return emscripten::val(emscripten::typed_memory_view(my_store.membership.size(), my_store.membership.data()));
    }
};

ClusterMultilevelResult cluster_multilevel(const BuildSnnGraphResult& graph, double resolution) {
    scran_graph_cluster::ClusterMultilevelOptions opt;
    opt.resolution = resolution;
    auto output = scran_graph_cluster::cluster_multilevel(graph.graph, graph.weights, opt);
    return ClusterMultilevelResult(std::move(output));
}

/**********************************/

class ClusterWalktrapResult {
    typedef scran_graph_cluster::ClusterWalktrapResults Store;

    Store my_store;
    std::size_t my_best = 0; 

public:
    ClusterWalktrapResult(Store s) : my_store(std::move(s)) {
        if (my_store.modularity.size()) {
            sanisizer::can_ptrdiff<I<decltype(my_store.modularity.begin())> >(my_store.modularity.size());
            my_best = std::max_element(my_store.modularity.begin(), my_store.modularity.end()) - my_store.modularity.begin();
        }
    }

public:
    JsFakeInt num_merge_steps() const {
        return int2js(my_store.merges.size());
    }

    double modularity(JsFakeInt i_raw) const {
        return my_store.modularity[js2int<std::size_t>(i_raw)];
    }

    double best_modularity() const {
        return my_store.modularity[my_best];
    }

    emscripten::val membership() const {
        return emscripten::val(emscripten::typed_memory_view(my_store.membership.size(), my_store.membership.data()));
    }
};

ClusterWalktrapResult cluster_walktrap(const BuildSnnGraphResult& graph, JsFakeInt steps_raw) {
    scran_graph_cluster::ClusterWalktrapOptions opt;
    opt.steps = js2int<igraph_int_t>(steps_raw);
    auto output = scran_graph_cluster::cluster_walktrap(graph.graph, graph.weights, opt);
    return ClusterWalktrapResult(std::move(output));
}

/**********************************/

class ClusterLeidenResult {
    typedef scran_graph_cluster::ClusterLeidenResults Store;

    Store my_store;

public:
    ClusterLeidenResult(Store s) : my_store(std::move(s)) {}

public:
    double quality() const {
        return my_store.quality;
    }

    emscripten::val membership() const {
        return emscripten::val(emscripten::typed_memory_view(my_store.membership.size(), my_store.membership.data()));
    }
};

ClusterLeidenResult cluster_leiden(const BuildSnnGraphResult& graph, double resolution, std::string objective) {
    scran_graph_cluster::ClusterLeidenOptions opt;
    opt.resolution = resolution;

    if (objective == "modularity") {
        opt.objective = IGRAPH_LEIDEN_OBJECTIVE_MODULARITY;
    } else if (objective == "cpm") {
        opt.objective = IGRAPH_LEIDEN_OBJECTIVE_CPM;
    } else if (objective == "er") {
        opt.objective = IGRAPH_LEIDEN_OBJECTIVE_ER;
    } else {
        throw std::runtime_error("unknown objective '" + objective + "'");
    }

    auto output = scran_graph_cluster::cluster_leiden(graph.graph, graph.weights, opt);
    return ClusterLeidenResult(std::move(output));
}

/**********************************/

EMSCRIPTEN_BINDINGS(cluster_graph) {
    emscripten::function("cluster_multilevel", &cluster_multilevel, emscripten::return_value_policy::take_ownership());

    emscripten::class_<ClusterMultilevelResult>("ClusterMultilevelResult")
        .function("num_levels", &ClusterMultilevelResult::num_levels, emscripten::return_value_policy::take_ownership())
        .function("best_level", &ClusterMultilevelResult::best_level, emscripten::return_value_policy::take_ownership())
        .function("modularity", &ClusterMultilevelResult::modularity, emscripten::return_value_policy::take_ownership())
        .function("membership", &ClusterMultilevelResult::membership, emscripten::return_value_policy::take_ownership())
        .function("best_modularity", &ClusterMultilevelResult::best_modularity, emscripten::return_value_policy::take_ownership())
        .function("best_membership", &ClusterMultilevelResult::best_membership, emscripten::return_value_policy::take_ownership())
        ;

    emscripten::function("cluster_walktrap", &cluster_walktrap, emscripten::return_value_policy::take_ownership());

    emscripten::class_<ClusterWalktrapResult>("ClusterWalktrapResult")
        .function("modularity", &ClusterWalktrapResult::modularity, emscripten::return_value_policy::take_ownership())
        .function("membership", &ClusterWalktrapResult::membership, emscripten::return_value_policy::take_ownership())
        .function("num_merge_steps", &ClusterWalktrapResult::num_merge_steps, emscripten::return_value_policy::take_ownership())
        ;

    emscripten::function("cluster_leiden", &cluster_leiden, emscripten::return_value_policy::take_ownership());

    emscripten::class_<ClusterLeidenResult>("ClusterLeidenResult")
        .function("quality", &ClusterLeidenResult::quality, emscripten::return_value_policy::take_ownership())
        .function("membership", &ClusterLeidenResult::membership, emscripten::return_value_policy::take_ownership())
        ;
}
